services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/update/UpdateExpression.java

```diff
@@ -0,0 +1,234 @@
+            return false;
+        }
+        if (setActions != null ? ! setActions.equals(that.setActions) : that.setActions != null) {
+            return false;
+        }
+        if (deleteActions != null ? ! deleteActions.equals(that.deleteActions) : that.deleteActions != null) {
+            return false;
+        }
+        return addActions != null ? addActions.equals(that.addActions) : that.addActions == null;
+    }
+
+    @Override
+    public int hashCode() {
+        int result = removeActions != null ? removeActions.hashCode() : 0;
+        result = 31 * result + (setActions != null ? setActions.hashCode() : 0);
+        result = 31 * result + (deleteActions != null ? deleteActions.hashCode() : 0);
+        result = 31 * result + (addActions != null ? addActions.hashCode() : 0);
+        return result;
+    }
+
+    /**
+     * A builder for {@link UpdateExpression}
+     */
+    public static final class Builder {
+
+        private Collection<RemoveUpdateAction> removeActions;
+        private Collection<SetUpdateAction> setActions;
+        private Collection<DeleteUpdateAction> deleteActions;
+        private Collection<AddUpdateAction> addActions;
+
+        private Builder() {
+        }
+
+        /**
+         * Add an action of type {@link RemoveUpdateAction}
+         */
+        public Builder addAction(RemoveUpdateAction action) {
+            if (this.removeActions == null) {
+                this.removeActions = new ArrayList<>();
+            }
+            removeActions.add(action);
+            return this;
+        }
+
+        /**
+         * Add an action of type {@link SetUpdateAction}
+         */
+        public Builder addAction(SetUpdateAction action) {
+            if (this.setActions == null) {
+                this.setActions = new ArrayList<>();
+            }
+            setActions.add(action);
+            return this;
+        }
+
+        /**
+         * Add an action of type {@link DeleteUpdateAction}
+         */
+        public Builder addAction(DeleteUpdateAction action) {
+            if (this.deleteActions == null) {
+                this.deleteActions = new ArrayList<>();
+            }
+            deleteActions.add(action);
+            return this;
+        }
+
+        /**
+         * Add an action of type {@link AddUpdateAction}
+         */
+        public Builder addAction(AddUpdateAction action) {
+            if (this.addActions == null) {
+                this.addActions = new ArrayList<>();
+            }
+            addActions.add(action);
+            return this;
+        }
+
+        /**
+         * Adds a collection of {@link UpdateAction} of any subtype to the builder. Calling this operation
+         * repeatedly will add the new collection to existing items and not overwrite them.
+         */
+        public Builder actions(Collection<UpdateAction> actions) {
+            addActions(actions);
```

Comment:
should this be called addActions? Elsewhere in the SDK, the noun form would override everything.

Reply:
Does it? I almost thought it was the other way around. Do you have any examples? I'm sure there's lots but it's hard to come up with a search term. 

Shouldn't we also look at the overall consistency between the action builders vs UpdateExpression?
UpdateExpression has 
- `addAction`
- `actions`

The action builders have
- `putExpressionValue` (and corresponding for expressionNames)
- `expressionValues`

The latter versions are taken from the existing `Expression` class and consistent with that naming. 
If we change `actions` to `addActions` then it's less consistent with `expressionValues`. 

What's really inconsistent is `addAction` vs `putExpressionValue`. Should we change it to `addExpressionValue`?

Reply:
`add` is for appending to lists/collections
`put` is for upserting to a map
no prefix overrides.

Prior art for add: https://github.com/aws/aws-sdk-java-v2/blob/master/http-client-spi/src/main/java/software/amazon/awssdk/http/SdkHttpRequest.java#L235-L253
Prior art for put: https://github.com/aws/aws-sdk-java-v2/blob/master/core/sdk-core/src/main/java/software/amazon/awssdk/core/client/config/ClientOverrideConfiguration.java#L528-L539

More information: https://github.com/aws/aws-sdk-java-v2/blob/master/docs/design/ClientConfiguration.md

Reply:
Thank you. I had forgotten about this design specification, which one should refer to..

Will align the plural method of adding a list of `action` with guidelines.